AlphaWallet/KeyManagement/JsonWalletAddresesStore.swift

```diff
@@ -0,0 +1,177 @@
+                return
+            }
+            _watchAddresses = newValue
+            storage.setData(data, forKey: Keys.watchAddresses)
+        }
+    }
+    private var _ethereumAddressesWithPrivateKeys: [String]?
+    var ethereumAddressesWithPrivateKeys: [String] {
+        get {
+            if let localValue = _ethereumAddressesWithPrivateKeys {
+                return localValue
+            } else if let value: [String] = storage.load(forKey: Keys.addressesWithPrivateKeys) {
+                return value
+            } else {
+                return []
+            }
+        }
+        set {
+            guard let data = try? JSONEncoder().encode(newValue) else {
+                return
+            }
+            _ethereumAddressesWithPrivateKeys = newValue
+            storage.setData(data, forKey: Keys.addressesWithPrivateKeys)
+        }
+    }
+
+    private var _ethereumAddressesWithSeed: [String]?
+    var ethereumAddressesWithSeed: [String] {
+        get {
+            if let localValue = _ethereumAddressesWithSeed {
+                return localValue
+            } else if let value: [String] = storage.load(forKey: Keys.addressesWithSeed) {
+                return value
+            } else {
+                return []
+            }
+        }
+        set {
+            guard let data = try? JSONEncoder().encode(newValue) else {
+                return
+            }
+            _ethereumAddressesWithSeed = newValue
+            storage.setData(data, forKey: Keys.addressesWithSeed)
+        }
+    }
+
+    private var _ethereumAddressesProtectedByUserPresence: [String]?
+    var ethereumAddressesProtectedByUserPresence: [String] {
+        get {
+            if let localValue = _ethereumAddressesProtectedByUserPresence {
+                return localValue
+            } else if let value: [String] = storage.load(forKey: Keys.addressesProtectedByUserPresence) {
+                return value
+            } else {
+                return []
+            }
+        }
+        set {
+            guard let data = try? JSONEncoder().encode(newValue) else {
+                return
+            }
+            _ethereumAddressesProtectedByUserPresence = newValue
+            storage.setData(data, forKey: Keys.addressesProtectedByUserPresence)
+        }
+    }
+}
+
+fileprivate struct Keys {
+    static let hasMigratedWalletAddressesToJsonFileStorage = "hasMigratedWalletAddressesToJsonFileStorage"
+}
+
+extension EtherKeystore {
+    static func migratedWalletAddressesStore(userDefaults: UserDefaults) -> WalletAddresesStoreType {
+        if Features.isJsonFileBasedStorageForWalletAddressesEnabled {
+            //NOTE: its quite important to remove test wallets right before fetching, otherwise tests will fails, especially Keystore related
+            JsonWalletAddresesStore.removeWalletsFolderForTests()
+
+            if !userDefaults.bool(forKey: Keys.hasMigratedWalletAddressesToJsonFileStorage) {
```

Comment:
Sorry, I didn't notice if this was here in the earlier version of this PR. But can we avoid using `UserDefaults` here (for JSON-based wallet storage functionality)? Because the reason why we switched was because we are having problems with `UserDefaults` at start up. If this flag returns a false when it shouldn't, we'll attempt to migrate again and wipe out the address lists.